Added support to pull chart from private repo

nraddy opened the pull request:
Signed-off-by: rj.natarajan <nraddy@gmail.com>

# What Are We Doing Here

Closes #60 

Added support to pull charts from private repositories. Used helm chat downloader to download the charts to local directory. If that fails, the flow falls back to using the chart from the configured directory when the controller was created or the previous version (if it was already downloaded).

The versioned chart is stored inside folder named with "base64 encoded versioned string" under temp directory. The change presumes that the private repo is added to the lostromos pod outside of the code, possibly during bootstrap.

## How to Verify

1. Check out this PR
2. make clean all
3. Profit

nicolerenee commented on Gopkg.toml:
```diff
@@ -24,7 +24,7 @@
 [[constraint]]
   name = "k8s.io/helm"
-  version = "2.8.0"
+  version = "2.9.0"
```
2.9.0 has some bugs that break new lines in yaml. 2.9.1 is out now which fixes this. We should skip 2.9.0 and go to the .1 release.

nraddy replied:
done!

nicolerenee commented on docs/helm.md:
```diff
@@ -64,3 +64,36 @@ spec:
+### Using custom repo for charts
+
+For Lostrómos to be able to pull the charts from a remote repository, the remote
+repository should be pre-configured in the Lostrómos pod (one option is to add it
```
This should state that it should be pre-configured in the tiller.

nraddy replied:
done!

nicolerenee commented on docs/helm.md:
````diff
@@ -50,7 +50,7 @@ variable name `resource`. The resource has the following fields available on it:
 
 ```yaml
-apiVersion: stable.nicolerenee.io/v1
+apiVersion: stable.foobar.io/v1
````
if we are going to change this it shouldn't be just some random url.

nraddy replied:
I usually don't put authors name in the code/doc unless needed, reason why I updated it when I added a new section with similar ymal block. (reverted back).

nicolerenee commented on helmctlr/remote_chart.go:
```diff
@@ -0,0 +1,96 @@
+	return ""
+}
+
+// Gets chart name and chart version from the passed chartRef, downloads the chart from repo,
```
this should start with getRemoteChart to match godoc format.

nraddy replied:
added it!

nicolerenee commented on docs/helm.md:
```diff
@@ -64,3 +64,36 @@ spec:
+  Name: Nemo
+  From: Finding Nemo
+  By: Disney
+  Chart: foonemo/helloworld:1.2r345
```
One thing we have always maintained for Lostrómos is the ability to run multiple instances against the same CRD, but have different outcomes. Hard coding this value to just use chart takes that away. 

To allow for this, we should use an annotation. It could default to `wpengine.com/lostromos.chart` but be changed via a config option.

dbrian replied:
I like this idea, it also feels cleaner not to have the chart definition passed as a value to the chart.

nraddy replied:
This change is in progress.

nraddy replied:
done!

nicolerenee commented on helmctlr/helm.go:
```diff
@@ -118,6 +118,16 @@ func (c Controller) installOrUpdate(r *unstructured.Unstructured) error {
+	// If chart entry is present in CR, then download the chart from repo to local dir
+	// on any error, fallback to local chart already present on the disk
+	chartRef := getChartRef(r)
+	if chartRef != "" {
```
Can this be tested?

nraddy replied:
done

nicolerenee commented on helmctlr/helm.go:
```diff
@@ -118,6 +118,16 @@ func (c Controller) installOrUpdate(r *unstructured.Unstructured) error {
+	chartRef := getChartRef(r)
+	if chartRef != "" {
+		if chartDir, chartErr := getRemoteChart(chartRef); chartErr == nil {
+			c.ChartDir = chartDir
```
I'm not sure I agree with this decision. If a remote chart is specified it is clear the user wants that chart to be deployed, not the local one. I don't think we should default to the local chart here. 

Either way, no matter which approach we take there should be logs to call out it failed.

nicolerenee commented on helmctlr/remote_chart.go:
```diff
@@ -0,0 +1,96 @@
+// Uses chart downloader to download the charts
+// 	- chart downloader uses the version passed, but if version is empty, pulls the latest version.
+// Prerequisite: Repo should have been initialized under HELM_HOME
+func getRemoteChart(chartRef string) (string, error) {
```
Can we add more tests to handle the error cases.

dbrian replied:
@nraddy Codecov report highlights the cases that need unit testing.

https://codecov.io/gh/wpengine/lostromos/pull/66/diff?src=pr&el=tree#diff-aGVsbWN0bHIvcmVtb3RlX2NoYXJ0Lmdv

nicolerenee commented:
Also, we have a very nice set of integration tests that were setup by @drush211. Can we add something to the integration test suite that ensures the remote repo functionality is working to avoid a regression in the future.

nraddy commented:
>> I also thing we should add metrics for the remote chart stuff, and potentially even call out how many releases are managed per chart.
@nicolerenee Could you be specific about what metrics need to be tracked here? I have added a metric to know every time remote chart is referenced, is that to be done per chart version?

nicolerenee commented:
The stuff you just pushed was what I had in mind for metrics at least as a starting point. It would be really cool to have a label with the remote chart in the metrics as well so that we can actually see the metrics for each remote chart. But I'll accept what we have here, the other is just a cherry on top.

nicolerenee commented:
@nraddy this is looking really good. The only real ask I have left is in regards to the integration tests.

nraddy commented:
@nicolerenee instructions are insufficient to run the integration tests locally. I suggest we create another issue to update the integration tests as this path is specific to remote chart only when added via annotations. The documentation on integration tests would also be updated along with that.

nraddy commented:
@nicolerenee @dbrian
Updated integration tests to include private repo

nraddy commented:
All changes are complete, waiting for approval - @nicolerenee , @dbrian

nraddy commented:
Travis passed until before the last commit:
https://travis-ci.org/wpengine/lostromos/builds/380864274
Last commit has only doc changes, the build is not running for some reason.

nicolerenee commented:
I've restarted the job to see if that makes any difference. I'll see if I can dig in and figure anything out once it completed.

nraddy commented:
Travis hasn't started for some reason again.

nicolerenee commented:
The build passed. For whatever reason it didn't report back to github. 

Can you squash this all down to a single commit per the contributing guidelines, then I think it should be ready to go.

nraddy commented:
Shall I do it in the same branch here? that would change the history, if that's okay.

nraddy commented:
Reading through the commits, I think that all of these can be merged into once single commit, as they are all relating to just one objective of adding private repo support - code change, testing, and doc update.
Please let me know, then I can squash merge all the changes as one and add this as the commit message:

```
Added support to pull chart from private repo
- Added chart entry to annotations in CRD
- Updated helm pkg version
- Updated tests, metrics and doc related to private repo

Signed-off-by: rj.natarajan <nraddy@gmail.com>
```

nicolerenee commented:
That works for me 👍. In the past we haven't worried about keeping the history from the branches, mostly we just want to keep clean commits to the master branch.

nraddy commented:
I think I need to push a code change for fixing travis-ci issue, so I am going to squash last two commits and re-push it to kick off travis build again. If you give your approval, I can merge the changes once travis build completes.

nraddy commented:
yup, that kicked off travis build. I think since the last commit had only doc change, the build did not kick-off, and got into a bad state here.

nraddy commented:
Just to be on the same page, I am (or whoever is merging the changes) going to squash commit the changes to master. 
(Squashing the changes in this branch will create confusion to the PR comments made here)

nicolerenee commented:
We can't do a squash commit on merge as that removes the DCO line which breaks the compliance for the repo. All the other PRs have had force push squashes on the branch and it's been okay. So you'll need to do that here as well.

nraddy commented:
done!

nraddy commented:
Did that, but the "update branch" appears still. I think I should update my fork's master branch to resolve this scenario.

nraddy commented:
all fixed now!
